File: api/src/main/java/com/castlefrog/agl/domains/hex/HexState.java
package com.castlefrog.agl.domains.hex;

import com.castlefrog.agl.State;

import java.io.Serializable;

public final class HexState implements State<HexState>, Serializable {
    private byte[][] bitBoards_;
    private byte boardSize_;
    private byte agentTurn_;
    private BoardState boardState_;

    public enum BoardState {
        EMPTY,
        FIRST_MOVE,
        SECOND_MOVE,
        OTHER
    }

    public static final int
        LOCATION_EMPTY = 0,
        LOCATION_BLACK = 1,
        LOCATION_WHITE = 2;

    public static final int
        TURN_BLACK = 0,
        TURN_WHITE = 1;

    public HexState(int boardSize,
                    byte[][] bitBoards,
                    int agentTurn,
                    BoardState boardState) {
        boardSize_ = (byte) boardSize;
        bitBoards_ = new byte[bitBoards.length][bitBoards[0].length];
        for (int i = 0; i < bitBoards.length; i += 1) {
            System.arraycopy(bitBoards[i], 0, bitBoards_[i], 0, bitBoards[0].length);
        }
        agentTurn_ = (byte) agentTurn;
        boardState_ = boardState;
    }

    public HexState copy() {
        return new HexState(boardSize_, bitBoards_, agentTurn_, boardState_);
    }

    public byte[][] getBitBoards() {
        return bitBoards_;
    }

    public byte[][] getLocations() {
        byte[][] locations = new byte[boardSize_][boardSize_];
        for (int i = 0; i < boardSize_; i += 1) {
            for (int j = 0; j < boardSize_; j += 1) {
                locations[i][j] = (byte) getLocation(i, j);
            }
        }
        return locations;
    }

    public int getLocation(int x, int y) {
        int bitLocation = y * boardSize_ + x;
        int byteLocation = bitLocation / Byte.SIZE;
        if ((bitBoards_[0][byteLocation] & (1 << bitLocation % Byte.SIZE)) != 0) {
            return LOCATION_BLACK;
        } else if ((bitBoards_[1][byteLocation] & ( 1 << bitLocation % Byte.SIZE)) != 0) {
            return  LOCATION_WHITE;
        } else {
            return LOCATION_EMPTY;
        }
    }

    public int getBoardSize() {
        return boardSize_;
    }

    public byte getAgentTurn() {
        return agentTurn_;
    }

    public BoardState getBoardState() {
        return boardState_;
    }

    public boolean isLocationEmpty(int x, int y) {
        int bitLocation = y * boardSize_ + x;
        int byteLocation = bitLocation / Byte.SIZE;
        return ((bitBoards_[0][byteLocation] | bitBoards_[1][byteLocation]) &
                (1 << bitLocation % Byte.SIZE)) == LOCATION_EMPTY;
    }

    public int getNPieces() {
        int nPieces = 0;
        for (int i = 0; i < bitBoards_[0].length; i += 1) {
            int value = bitBoards_[0][i] | bitBoards_[1][i];
            for (int j = 0; j < Byte.SIZE; j += 1) {
                if ((value & 0b1) != 0) {
                    nPieces += 1;
                }
                value = value >>> 1;
            }
        }
        return nPieces;
    }

    public void setLocation(int x, int y, int value) {
        int bitLocation = y * boardSize_ + x;
        int byteLocation = bitLocation / Byte.SIZE;
        int byteShift = bitLocation % Byte.SIZE;
        if (value == LOCATION_EMPTY) {
            bitBoards_[0][byteLocation] &= ((1 << byteShift) ^ 0xff);
            bitBoards_[1][byteLocation] &= ((1 << byteShift) ^ 0xff);
        } else if (value == LOCATION_BLACK) {
            bitBoards_[0][byteLocation] |= 1 << byteShift;
            bitBoards_[1][byteLocation] &= ((1 << byteShift) ^ 0xff);
        } else if (value == LOCATION_WHITE) {
            bitBoards_[0][byteLocation] &= ((1 << byteShift) ^ 0xff);
            bitBoards_[1][byteLocation] |= 1 << byteShift;
        }
    }

    public void setAgentTurn(int agentTurn) {
        agentTurn_ = (byte) agentTurn;
    }

    public void setBoardState(BoardState boardState) {
        boardState_ = boardState;
    }

    @Override
    public int hashCode() {
        int hashCode = 17;
        hashCode = hashCode * 13 + boardSize_;
        hashCode = hashCode * 23 + agentTurn_;
        for (byte[] row : bitBoards_) {
            for (byte location : row) {
                hashCode = 31 * hashCode + location;
            }
        }
        return hashCode;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof HexState)) {
            return false;
        }
        HexState state = (HexState) object;
        byte[][] bitBoards = state.getBitBoards();
        for (int i = 0; i < bitBoards_.length; i += 1) {
            for (int j = 0; j < bitBoards_.length; j += 1) {
                if (bitBoards_[i][j] != bitBoards[i][j]) {
                    return false;
                }
            }
        }
        return agentTurn_ == state.getAgentTurn() &&
               boardSize_ == state.getBoardSize();
    }

    @Override
    public String toString() {
        StringBuilder output = new StringBuilder();
        for (int i = boardSize_ - 1; i >= 0; i -= 1) {
            for (int j = i; j < boardSize_ - 1; j += 1) {
                output.append(" ");
            }
            for (int j = 0; j < boardSize_; j += 1) {
                int location = getLocation(j, i);
                if (location == LOCATION_BLACK) {
                    output.append("X ");
                } else if (location == LOCATION_WHITE) {
                    output.append("O ");
                } else {
                    output.append("- ");
                }
            }
            output.append("\n");
        }
        return output.toString();
    }
}

File: api/src/main/java/com/castlefrog/agl/domains/hex/HexState.kt
package com.castlefrog.agl.domains.hex

import com.castlefrog.agl.State

import java.io.Serializable

data class HexState(val boardSize: Int,
               val bitBoards: Array<ByteArray>,
               var agentTurn: Byte) : State<HexState>, Serializable {

    companion object {
        val LOCATION_EMPTY = 0
        val LOCATION_BLACK = 1
        val LOCATION_WHITE = 2

        val TURN_BLACK: Byte = 0
        val TURN_WHITE: Byte = 1
    }

    override fun copy(): HexState {
        return HexState(boardSize, bitBoards, agentTurn)
    }

    val locations: Array<ByteArray>
        get() {
            val locations = Array(boardSize) { ByteArray(boardSize) }
            var i = 0
            while (i < boardSize) {
                var j = 0
                while (j < boardSize) {
                    locations[i][j] = getLocation(i, j).toByte()
                    j += 1
                }
                i += 1
            }
            return locations
        }

    fun getLocation(x: Int, y: Int): Int {
        checkLocationArgs(x, y)
        val bitLocation = y * boardSize + x
        val byteLocation = bitLocation / java.lang.Byte.SIZE
        if (bitBoards[0][byteLocation].toInt() and (1 shl bitLocation % java.lang.Byte.SIZE) != 0) {
            return LOCATION_BLACK
        } else if (bitBoards[1][byteLocation].toInt() and (1 shl bitLocation % java.lang.Byte.SIZE) != 0) {
            return LOCATION_WHITE
        } else {
            return LOCATION_EMPTY
        }
    }

    fun isLocationEmpty(x: Int, y: Int): Boolean {
        checkLocationArgs(x, y)
        val bitLocation = y * boardSize + x
        val byteLocation = bitLocation / java.lang.Byte.SIZE
        return bitBoards[0][byteLocation].toInt() or bitBoards[1][byteLocation].toInt() and (1 shl bitLocation % java.lang.Byte.SIZE) == LOCATION_EMPTY
    }

    fun isLocationOnBoard(x: Int, y: Int): Boolean {
        return x >= 0 && x < boardSize && y >= 0 && y < boardSize
    }

    val nPieces: Int
        get() {
            var nPieces = 0
            for (i in 0..bitBoards[0].size - 1) {
                var value = bitBoards[0][i].toInt() or bitBoards[1][i].toInt()
                for (j in 0..java.lang.Byte.SIZE - 1) {
                    if (value and 1 != 0) {
                        nPieces += 1
                    }
                    value = value.ushr(1)
                }
            }
            return nPieces
        }

    fun setLocation(x: Int, y: Int, value: Int) {
        checkLocationArgs(x, y)
        val bitLocation = y * boardSize + x
        val byteLocation = bitLocation / java.lang.Byte.SIZE
        val byteShift = bitLocation % java.lang.Byte.SIZE
        if (value == LOCATION_EMPTY) {
            bitBoards[0][byteLocation] = (bitBoards[0][byteLocation].toInt() and (1 shl byteShift xor 0xff)).toByte()
            bitBoards[1][byteLocation] = (bitBoards[1][byteLocation].toInt() and (1 shl byteShift xor 0xff)).toByte()
        } else if (value == LOCATION_BLACK) {
            bitBoards[0][byteLocation] = (bitBoards[0][byteLocation].toInt() or (1 shl byteShift)).toByte()
            bitBoards[1][byteLocation] = (bitBoards[1][byteLocation].toInt() and (1 shl byteShift xor 0xff)).toByte()
        } else if (value == LOCATION_WHITE) {
            bitBoards[0][byteLocation] = (bitBoards[0][byteLocation].toInt() and (1 shl byteShift xor 0xff)).toByte()
            bitBoards[1][byteLocation] = (bitBoards[1][byteLocation].toInt() or (1 shl byteShift)).toByte()
        }
    }

    private fun checkLocationArgs(x: Int, y: Int) {
        if (!isLocationOnBoard(x, y))
            throw IllegalArgumentException("(x=$x,y=$y) out of bounds")
    }

    override fun equals(other: Any?): Boolean {
        if (other !is HexState) {
            return false
        }
        if (bitBoards.size != other.bitBoards.size) {
            return false
        }
        for (i in bitBoards.indices) {
            if (bitBoards[i].size != other.bitBoards[i].size) {
                return false
            }
            for (j in bitBoards[i].indices) {
                if (other.bitBoards[i][j] != bitBoards[i][j]) {
                    return false
                }
            }
        }
        return other.boardSize == boardSize && other.agentTurn == agentTurn
    }

    override fun toString(): String {
        val output = StringBuilder()
        output.append("turn = ").append(agentTurn).append("\n")
        for (i in boardSize - 1 downTo 0) {
            for (j in i..boardSize - 2) {
                output.append(" ")
            }
            for (j in 0..boardSize - 1) {
                val location = getLocation(j, i)
                if (location == LOCATION_BLACK) {
                    output.append("X")
                } else if (location == LOCATION_WHITE) {
                    output.append("O")
                } else {
                    output.append("-")
                }
                if (j != boardSize - 1) {
                    output.append(" ")
                }
            }
            if (i != 0) {
                output.append("\n")
            }
        }
        return output.toString()
    }
}
